src/libsystemd-network/sd-dhcp-lease.c

```diff
@@ -559,6 +572,218 @@ static int lease_parse_sip_server(const uint8_t *option, size_t len, struct in_a
         return lease_parse_in_addrs(option + 1, len - 1, ret, n_ret);
 }
 
+//FIXME: this is wonky
+static int lease_parse_dns_name(const uint8_t *option, size_t len, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        char *p;
+
+        if (len == 0)
+                return -EINVAL;
+
+        if (strnlen((const char*) option+1, len-1) != len-2)
+                return -EINVAL;
+
+        name = new0(char, len);
+        p = stpncpy(name, (const char*) option + 1, len - 1);
+        if (p != name + len - 2)
+                return -EINVAL;
+
+        size_t n = *option;
+        while (n < len - 1) {
+                uint8_t tmp = name[n];
+                name[n] = '.';
+                n += tmp + 1;
+        }
+        if (n != len-1 || name[n] != '\0')
+                return -EINVAL;
+
+        *ret = TAKE_PTR(name);
+        return 0;
+}
+
+static int lease_parse_dnr_params(const uint8_t *option, size_t len, ResolverData *resolver) {
+        size_t offset = 0;
+
+        assert(resolver);
+
+        if (len < 4)
+                return -EINVAL;
+
+        DNSALPNFlags transports = 0;
+        uint16_t port = 0;
+        _cleanup_free_ char *dohpath = NULL;
+        bool alpn = false;
+
+        while (offset < len) {
+                uint16_t key = unaligned_read_be16(&option[offset]);
```

Comment:
doesn't matter, but I would always write `option + offset` rather than `&option[offset]`...

Reply:
I continued to use the `&option[offset]` format, just because it reads a little easier in my head ("give me a pointer to offset in buffer"). I don't mind converting it now that the feature is mostly implemented, if you'd prefer.

Reply:
Ok, I converted these to '+' pointer arithmetic.